### ISSUE TITLE
Base class function signature compatibility checks overly strict

### ISSUE THREAD

**Issue by ambv** [python/mypy#2619]
Actual example of this code can be found in Python 3.6's enum with `Flag` and `IntFlag`.

**Minimal repro:**:
```py3
from typing import TypeVar

_T = TypeVar('_T')


class Base:
    def __or__(self: _T, other: _T) -> _T: ...
    def __and__(self: _T, other: _T) -> _T: ...
    def __xor__(self: _T, other: _T) -> _T: ...


class Other(int, Base):  # <-- errors here
    def __or__(self: _T, other: Union[int, _T]) -> _T: ...  # type: ignore
    def __and__(self: _T, other: Union[int, _T]) -> _T: ...  # type: ignore
    def __xor__(self: _T, other: Union[int, _T]) -> _T: ...  # type: ignore
```

**Expected behavior**: no errors.
**Actual behavior**:
```
f:11: error: Definition of "__and__" in base class "int" is incompatible with definition in base class "Base"
f:11: error: Definition of "__or__" in base class "int" is incompatible with definition in base class "Base"
f:11: error: Definition of "__xor__" in base class "int" is incompatible with definition in base class "Base"
```

The errors would be fine if `Other` didn't redefine the methods itself. In this case the incompatibility in base classes is irrelevant.

**Comment by JukkaL** [comment]
Keywords to help searching: self type, union type

**Comment by jingw** [comment]
Just ran into this at Dropbox.

Slightly more distilled repro:
```python
class A(object):
    def x(self):
        # type: () -> A
        return self

class B(object):
    def x(self):
        # type: () -> B
        return self

class C(A, B):
    def x(self):
        # type: () -> C
        return self

test.py:11: error: Definition of "x" in base class "A" is incompatible with definition in base class "B"
```

And equivalent compilable Java code:
```java
interface A {
    A x();
}
interface B {
    B x();
}
class C {
    C x() {
        return this;
    }
}
```

I presume mypy already has checks that the C.x is compatible with A.x and B.x. Is it possible to resolve this issue by doing something along the lines of removing the check that A.x and B.x are compatible in the case where x is overridden? (Note, I don't really know what I'm talking about.)
